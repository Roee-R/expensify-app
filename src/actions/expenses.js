import uuid from 'uuid'; // generate id

 // Add expense 
 export const addExpense = ( 
    { 
        description='', note='', amount=0, createdAt=0 // USING OBJECT DESTRUCTURIN
    } ={})=>(
        { // return OBJECT
            type: 'ADD_EXPENSE',
            expenses: {
                id: uuid(),
                description,
                note,
                amount,
                createdAt
            }
        }   
    )

// Remove expanse
export const removeExpense = ({id})=>(
        { // return OBJECT
            type: 'REMOVE_EXPENSE',
            id
        }   
    )
    
// Edit expanse

export const editExpense = (id, updates) =>({
    type: 'EDIT_EXPENSE',
    id,
    updates
})