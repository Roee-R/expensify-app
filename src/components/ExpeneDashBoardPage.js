import React from 'react';

import ExpanseList from './ExpenseList';

const AddexpeneDashBoardPage = () =>{
    return(
    <div>
        <h1>Wellcome to my DashBoard component</h1>
        <ExpanseList
        />
    </div>    
)
    }
export default AddexpeneDashBoardPage;

