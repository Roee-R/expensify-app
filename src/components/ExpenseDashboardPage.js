import React from 'react';
import { connect } from 'react-redux';

import ExpanseList from './ExpenseList';
import ExpensesSummary from './ExpensesSummary';

import selectdExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expenses-total';
import ExpenseListFilter from './ExpenseListFilter'; // filtering by name

export const ExpenseDashboardPage = (props) =>{
    return(
    <div>
        <ExpensesSummary 
            size={props.expensesSize}
            sum={props.sum}/>
        <ExpenseListFilter />    
        <ExpanseList
        />
    </div>    
        )
}


const mapStateToProps = (state) =>{
    const visibleExpenses = selectdExpenses(state.expenses, state.filters)
    const sum = selectExpensesTotal(visibleExpenses);
    return {
        expensesSize: visibleExpenses.length,
        sum
    }
}


export default connect(mapStateToProps)(ExpenseDashboardPage)
    
