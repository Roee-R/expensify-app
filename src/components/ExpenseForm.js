import React from 'react';

import {SingleDatePicker} from 'react-dates';
import 'react-dates/initialize';
import moment from 'moment';

export default class ExpenseForm extends React.PureComponent {
    constructor(props){
        super(props)
        this.state = {
            description: props.expense?props.expense.description:'',
            note: props.expense?props.expense.note:'',
            amount: props.expense?(props.expense.amount/100).toString():'',
            createdAt: props.expense?moment(props.expense.createdAt):moment(),
            calendarFocused: false,
            error: ''
        }
    }

    onDescriptionChange = (e)=>{
        const description = e.target.value;
        this.setState(()=>({
            description
        }))
    }

    onNoteChange = (e)=>{
        const note = e.target.value;
        this.setState(()=>({
            note
        }))
    }

    onAmountChanage = (e) =>{
        const amount = e.target.value;
        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){ //rejex for having max 2 decimal point
            this.setState(()=>({
                amount
            }))
        }
    }

    onSubmit = (e)=>{
        e.preventDefault();
        if(this.state.description==='' || this.state.amount===''){
            this.setState(()=>({error: 'Please type decripction and expense amount'}))
        }
        else{
            this.setState(()=>({error: ''}))
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10)*100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            })
        }
    }
    render(){
        return (
            <form onSubmit={this.onSubmit}>
                {this.state.error && <p>{this.state.error}</p>}
                <input 
                type="text"
                placeholder="Expense title"
                value={this.state.description}
                onChange={this.onDescriptionChange}
                autoFocus  />
                <input 
                type="number"
                placeholder="amount" 
                value={this.state.amount}
                onChange={this.onAmountChanage} />
                <SingleDatePicker
                date={this.state.createdAt} // momentPropTypes.momentObj or null
                onDateChange={createdAt => createdAt && this.setState({ createdAt })} // PropTypes.func.isRequired
                focused={this.state.calendarFocused} // PropTypes.bool
                onFocusChange={({ focused }) => this.setState({ calendarFocused:focused })} // PropTypes.func.isRequired
                numberOfMonths= {1}
                isOutsideRange={()=>false}
                // id="your_unique_id" // PropTypes.string.isRequired,
                />
                <textarea
                placeholder="Add a note for your expense"
                onChange={this.onNoteChange}
                value={this.state.note}
                >
                </textarea>
                <button>Add expense</button>
            </form>
        )
    }
}

export const pureComponentAvailable = true;
