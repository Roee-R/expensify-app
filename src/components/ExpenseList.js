import React from 'react';
import { connect } from 'react-redux';

import getVisibleExpanse from '../selectors/expenses' // return sorted array
import ExpenseListItem from './ExpenseListItem'; // Return item from the list
import ExpenseListFilter from './ExpenseListFilter'; // filtering by name

export const ExpenseList = (props) =>(
    <div>
        <ExpenseListFilter />
        {props.expenses.length>0 
        ? (props.expenses.map((expense)=>{
            return (<ExpenseListItem key={expense.id} 
            index={props.expenses.indexOf(expense)}
            {...expense} 
            />)
        }))
        : (<h3>No expenses</h3>)
    }
        
    </div>
)
const mapStateToProps = (state) =>{
    return {
        expenses: getVisibleExpanse(state.expenses, state.filters)
    };
};

export default connect(mapStateToProps)(ExpenseList);