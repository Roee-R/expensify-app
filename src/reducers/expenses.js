// expenses Reduser

const expensesReduserDefaultState = []; // default array state

// Create expensesReduser Reducer
const expensesReduser = (state = expensesReduserDefaultState, action)=>{ // action is the extra propeties that we delivar via reducer object TYPES functions like editExpense
    switch(action.type){
        case 'ADD_EXPENSE':{
            return [...state, action.expenses]; // ...state is new syntax for arrays "spearign array"
        }
        case 'REMOVE_EXPENSE':{
            return [...state].filter(( {id} )=>id!==action.id);
        }
        case 'EDIT_EXPENSE':{
            return state.map(( expense )=>{
                if(expense.id===action.id){
                    return {
                        ...expense, ...action.updates //{...expense} is new syntax for arrays "spearign object"
                    };
                }else{
                    return expense
                }
            })
        }
        default:{
            return state;
        }
    }
}

export default expensesReduser;