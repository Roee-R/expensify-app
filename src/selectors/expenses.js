import moment from 'moment'

// sort and filter function
const getVisibleExpense = (expenses,{text, sortBy, startDate, endDate})=>{ // destructuring

    return expenses.filter((expense)=>{ // filter return just values with number bitween start date andend date
            const momentCreatedAt= moment(expense.createdAt)
            const startDateMatch=startDate?startDate.isSameOrBefore(momentCreatedAt, 'day'):true
            const endDateMatch=endDate?momentCreatedAt.isSameOrBefore(endDate, 'day'):true   
            const textMatch = expense.description.toLowerCase().includes(text.toLowerCase()); 

            return startDateMatch && endDateMatch && textMatch;
        }).sort((a,b)=>{ // sorting by amount or by date
            if(sortBy.toLowerCase()==='amount')
                return b.amount-a.amount;
            else if(sortBy.toLowerCase()==='date'){
                return b.createdAt-a.createdAt; 
            }   
        })  
}

export default getVisibleExpense;