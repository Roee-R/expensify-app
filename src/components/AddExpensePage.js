import React from 'react';
import { connect } from 'react-redux'; // import the connect the method of react-redux

import ExpenseForm from './ExpenseForm';
import {addExpense} from '../actions/expenses'

export class AddExpensePage extends React.Component{
    onSubmit=(expenses)=>{
        this.props.addExpense(expenses);
        this.props.history.push('/');
    }
    render(){
        return (
            <div>
                <h1>Wellcome from my add expense component</h1>
                <ExpenseForm 
                onSubmit={this.onSubmit}
                />
            </div>   
        )
    }
}

const mapDispatchToProps = (dispatch)=>({ // function that transfer dispatch to components
    addExpense: (expenses)=>dispatch(addExpense(expenses)) // use abtract dispatch to use on the web and on the tests
})

export default connect(undefined,mapDispatchToProps)(AddExpensePage); // connect the store to the component (to update store data) and link the mapStateToProps

