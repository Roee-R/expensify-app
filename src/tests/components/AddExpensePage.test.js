import React from 'react';
import {shallow} from 'enzyme';
import { AddExpensePage } from "../../components/AddExpensePage";

import expenses from '../fixtures/expenses';

let addExpenseSpy, historySpy, wraper;

beforeEach(()=>{
    addExpenseSpy = jest.fn(); // set spy fucntion
    historySpy= {push: jest.fn()};
    wraper = shallow(<AddExpensePage 
        history={historySpy}
        addExpense={addExpenseSpy}/>)
})

describe('Expense page tests', ()=>{
    test('should render AddExpensePage correctly', ()=>{
        expect(wraper).toMatchSnapshot();
    })

    test('should handle on submit', ()=>{
        wraper.find('ExpenseForm').prop('onSubmit')(expenses[1]);
        expect(historySpy.push).lastCalledWith('/');
        expect(addExpenseSpy).toHaveBeenCalledWith(expenses[1]);    
    })
})