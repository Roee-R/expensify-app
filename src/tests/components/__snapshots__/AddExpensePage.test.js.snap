// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expense page tests should render AddExpensePage correctly 1`] = `
<div>
  <h1>
    Wellcome from my add expense component
  </h1>
  <ExpenseForm
    onSubmit={[Function]}
  />
</div>
`;
